### ISSUE TITLE
'TypeError: argument of type 'File' is not iterable' when passing files as executable name

### ISSUE THREAD

**Issue by thatlittlegit** [mesonbuild/meson#11460]
**Describe the bug**
Passing a `files` object to `executable` as the name can cause a `TypeError` instead of a useful error message.

**To reproduce**
Inside the same directory as the meson.build, create a subdirectory 'a' containing 'b.c'. In meson.build:
```meson
project('name', 'c')
src = files('a/b.c')
exe = executable(src)
```
Then run `meson setup build` or similar.

**Expected behavior**
A Meson (non-Python) error telling you that the files object is actually the second parameter, and that the first one must be a string. (Maybe also one that there's too few positional parameters?)

**Actual behavior**
<details>
<summary>Output log</summary>
<pre>
user@host:/tmp$ mkdir meson-bug
user@host:/tmp$ cd meson-bug/
user@host:/tmp/meson-bug$ cat >>meson.build
project('name', 'c')
src = files('a/b.c')        
exe = executable(src)
user@host:/tmp/meson-bug$ mkdir a
user@host:/tmp/meson-bug$ touch a/b.c
user@host:/tmp/meson-bug$ meson setup build
The Meson build system
Version: 1.0.1
Source dir: /tmp/meson-bug
Build dir: /tmp/meson-bug/build
Build type: native build
Project name: name
Project version: undefined
C compiler for the host machine: ccache cc (gcc 12.2.0 "cc (Debian 12.2.0-14) 12.2.0")
C linker for the host machine: cc ld.bfd 2.40
Host machine cpu family: x86_64
Host machine cpu: x86_64
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/mesonbuild/mesonmain.py", line 194, in run
    return options.run_func(options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/mesonbuild/msetup.py", line 310, in run
    app.generate()
  File "/usr/lib/python3/dist-packages/mesonbuild/msetup.py", line 187, in generate
    self._generate(env)
  File "/usr/lib/python3/dist-packages/mesonbuild/msetup.py", line 231, in _generate
    intr.run()
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreter/interpreter.py", line 2875, in run
    super().run()
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 147, in run
    self.evaluate_codeblock(self.ast, start=1)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 172, in evaluate_codeblock
    raise e
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 165, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 180, in evaluate_statement
    self.assignment(cur)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 573, in assignment
    value = self.evaluate_statement(node.value)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 178, in evaluate_statement
    return self.function_call(cur)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 460, in function_call
    res = func(node, func_args, kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/decorators.py", line 784, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/decorators.py", line 784, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/decorators.py", line 784, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/decorators.py", line 127, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreter/interpreter.py", line 1720, in func_executable
    return self.build_target(node, args, kwargs, build.Executable)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreter/interpreter.py", line 3156, in build_target
    target = targetclass(name, self.subdir, self.subproject, for_machine, srcs, struct, objs,
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/mesonbuild/build.py", line 1792, in __init__
    super().__init__(name, subdir, subproject, for_machine, sources, structured_sources, objects,
  File "/usr/lib/python3/dist-packages/mesonbuild/build.py", line 689, in __init__
    super().__init__(name, subdir, subproject, True, for_machine, environment)
  File "<string>", line 9, in __init__
  File "/usr/lib/python3/dist-packages/mesonbuild/build.py", line 517, in __post_init__
    if has_path_sep(self.name):
       ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/mesonbuild/utils/universal.py", line 1074, in has_path_sep
    if each in name:
       ^^^^^^^^^^^^
TypeError: argument of type 'File' is not iterable

meson.build:3:0: ERROR: Unhandled python exception

&nbsp;&nbsp;&nbsp;&nbsp;This is a Meson bug and should be reported!
</pre>
</details>

**system parameters**
* _Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)?_ Just a plain native build.
* _what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.)_ Debian sid, updated earlier today
* _what Python version are you using_ 3.11.2
* _what `meson --version`_ 1.0.1
* _what `ninja --version` if it's a Ninja build_ 1.11.1

**Comment by xclaesse** [comment]
I believe this has already been fixed by https://github.com/mesonbuild/meson/pull/11293

**Comment by eli-schwartz** [comment]
Indeed, the fix for this will be in meson 1.1.0. The error is:
```
meson.build:3:0: ERROR: executable argument 1 was of type "File" but should have been "str"
```

(There is a separate warning if you pass valid typechecking and forget to list any sources in the executable.)
